compose/ui/ui/src/webCommonW3C/kotlin/androidx/compose/ui/platform/accessibility/ComposeWebSemanticsListener.kt

```diff
@@ -0,0 +1,363 @@
+import kotlinx.coroutines.flow.merge
+import kotlinx.coroutines.flow.receiveAsFlow
+import kotlinx.coroutines.flow.sample
+import kotlinx.coroutines.launch
+import org.w3c.dom.HTMLElement
+
+internal class ComposeWebSemanticsListener(
+    val coroutineScope: CoroutineScope,
+    val webSemanticsRoot: HTMLElement,
+) : PlatformContext.SemanticsOwnerListener {
+
+    private val invalidationChannel =
+        Channel<Unit>(1, onBufferOverflow = BufferOverflow.DROP_LATEST)
+
+    init {
+        coroutineScope.launch {
+            var lastSyncTimeMs = currentTimeMillis()
+            val invalidationFlow = invalidationChannel.receiveAsFlow()
+
+            @OptIn(FlowPreview::class)
+            merge(
+                invalidationFlow.sample(1000),
+                invalidationFlow.debounce(100)
+            ).collect {
+                val currentTime = currentTimeMillis()
+                if (currentTime - lastSyncTimeMs >= 100) {
+                    lastSyncTimeMs = currentTime
+                    syncSemanticsWithWebA11Y()
+                }
+            }
+        }
+    }
+
+    private var semanticsOwner: SemanticsOwner? = null
+
+    override fun onSemanticsOwnerAppended(semanticsOwner: SemanticsOwner) {
+        this.semanticsOwner = semanticsOwner
+    }
+
+    override fun onSemanticsOwnerRemoved(semanticsOwner: SemanticsOwner) {
+        if (semanticsOwner == this.semanticsOwner) {
+            this.semanticsOwner = null
+        }
+    }
+
+    override fun onSemanticsChange(semanticsOwner: SemanticsOwner) {
+        invalidationChannel.trySend(Unit)
+    }
+
+    override fun onLayoutChange(
+        semanticsOwner: SemanticsOwner, semanticsNodeId: Int
+    ) {
+        invalidationChannel.trySend(Unit)
+    }
+
+    private val dfsDeque = ArrayDeque<SemanticsNode>()
+
+    private val nodes = MutableScatterMap<Int, SemanticsNode>()
+    private val nodeToParent = MutableScatterMap<Int, Int>()
+    private val webNodes = MutableScatterMap<Int, HTMLElement>()
+
+
+    private fun syncSemanticsWithWebA11Y() {
+        fun SemanticsNode.isValid() = layoutNode.let { it.isPlaced && it.isAttached }
+
+        val root = semanticsOwner?.rootSemanticsNode ?: return
+
+        if (root.isValid()) {
+            dfsDeque.addLast(root)
+        }
+
+
+        val allIds = mutableSetOf<Int>()
+
+        val rootPosition = webSemanticsRoot.getBoundingClientRect().let {
+            Offset(it.left.toFloat(), it.top.toFloat())
+        }
+
+        while (!dfsDeque.isEmpty()) {
+            val node = dfsDeque.removeLast()
+            val currentId = node.id
+            allIds.add(currentId)
+
+            val children = node.replacedChildren.asReversed()
+            dfsDeque.addAll(children)
+            children.forEach { it -> nodeToParent[it.id] = currentId }
+
+            val htmlNode = if (nodes[currentId] != null) {
+                nodes[currentId] = node
+                val htmlNode = webNodes[currentId]!!
```

Comment:
instead of `!!` I'd use `?: error("description")`

Reply:
fixed